File: src/main/java/lrusso96/simplebiblio/core/Provider.java
package lrusso96.simplebiblio.core;

import lrusso96.simplebiblio.exceptions.BiblioException;

import java.util.List;
import java.util.logging.Level;

public abstract class Provider {

    public static final String LIBGEN = "Library Genesis";
    public static final String FEEDBOOKS = "Feedbooks";
    public static final String STANDARD_EBOOKS = "Standard Ebooks";

    protected String name;
    private SimpleBiblio biblio;

    public Provider(String name, SimpleBiblio biblio) {
        this.name = name;
        this.biblio = biblio;
    }

    public List<Ebook> search(String query) throws BiblioException {
        int cnt = 0;
        while (true)
            try {
                return doSearch(query);
            } catch (BiblioException ex) {
                if (++cnt == biblio.getMaxTries())
                    throw ex;
            }
    }

    protected abstract List<Ebook> doSearch(String query) throws BiblioException;

    public List<Ebook> getRecent() throws BiblioException {
        int cnt = 0;
        while (true)
            try {
                return doGetRecent();
            } catch (BiblioException ex) {
                if (++cnt == biblio.getMaxTries())
                    throw ex;
            }
    }

    protected abstract List<Ebook> doGetRecent() throws BiblioException;

    public List<Ebook> getPopular() throws BiblioException {
        int cnt = 0;
        while (true)
            try {
                return doGetPopular();
            } catch (BiblioException ex) {
                if (++cnt == biblio.getMaxTries())
                    throw ex;
            }
    }

    protected abstract List<Ebook> doGetPopular() throws BiblioException;

    public String getName() {
        return name;
    }

    protected void log(Level level, String str) {
        if (biblio.logger != null)
            biblio.logger.log(level, str);
    }
}

File: src/main/java/lrusso96/simplebiblio/core/Provider.kt
package lrusso96.simplebiblio.core

import lrusso96.simplebiblio.exceptions.BiblioException
import mu.KotlinLogging

private val logger = KotlinLogging.logger {}

abstract class Provider(val name: String, private val maxTries: Int = 3) {

    suspend fun search(query: String): List<Ebook> {
        var cnt = 0
        while (true) try {
            return doSearch(query)
        } catch (ex: BiblioException) {
            if (++cnt == maxTries) {
                logger.error { "$name failed to retrieve result for $maxTries time(s)" }
                return ArrayList()
            }
        }
    }

    @Throws(BiblioException::class)
    protected abstract suspend fun doSearch(query: String): List<Ebook>

    suspend fun getRecent(): List<Ebook> {
        var cnt = 0
        while (true) try {
            return doGetRecent()
        } catch (ex: BiblioException) {
            if (++cnt == maxTries) {
                logger.error { "$name failed to retrieve result for $maxTries time(s)" }
                return ArrayList()
            }
        }
    }

    @Throws(BiblioException::class)
    protected abstract suspend fun doGetRecent(): List<Ebook>

    suspend fun getPopular(): List<Ebook> {
        var cnt = 0
        while (true) try {
            return doGetPopular()
        } catch (ex: BiblioException) {
            if (++cnt == maxTries) {
                logger.error { "$name failed to retrieve result for $maxTries time(s)" }
                return ArrayList()
            }
        }
    }

    @Throws(BiblioException::class)
    protected abstract suspend fun doGetPopular(): List<Ebook>

    companion object {
        const val LIBGEN = "Library Genesis"
        const val FEEDBOOKS = "Feedbooks"
        const val STANDARD_EBOOKS = "Standard Ebooks"
    }
}